d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperServer.java

```diff
@@ -47,18 +49,32 @@
 {
   private static final Logger _log = LoggerFactory.getLogger(ZooKeeperServer.class);
 
+  // we set a default NoopIndisAnnouncer so that even if the user doesn't provide an announcer the class
+  // functions without throwing a NPE when marking up or down
+  private IndisAnnouncer _indisAnnouncer = new NoopIndisAnnouncer();
+
   private volatile ZooKeeperEphemeralStore<UriProperties> _store;
   private D2ServiceDiscoveryEventHelper _eventHelper = null;
 
   public ZooKeeperServer()
   {
   }
 
+  public ZooKeeperServer(IndisAnnouncer indisAnnouncer) {
```

Comment:
how could we construct this class without initializing the ZooKeeperEphemeralStore?

Reply:
In the `D2ServerFactory` in container we use the no-argument `ZookeeperServer` constructor, which is where we will be we initializing the IndisAnnouncer implementation. I think the reason it works is because somewhere in the code they use `setStore` to assign the store before the D2Server actually starts